src/libopensc/card-openpgp.c

```diff
@@ -2082,6 +2194,166 @@ pgp_put_data(sc_card_t *card, unsigned int tag, const u8 *buf, size_t buf_len)
 	LOG_FUNC_RETURN(card->ctx, (int)buf_len);
 }
 
+#ifdef ENABLE_OPENSSL
+/* KDF_ITERSALTED_S2K algorithm, a simplified version of openpgp_s2k from libgcrypt[1]
+ * that supports one-pass only and outputs the same key length as the hash length of
+ * the specified hash algorithm.
+ * [1]: https://github.com/gpg/libgcrypt/blob/libgcrypt-1.11.0/cipher/kdf.c#L32-L109
+ */
+static int
+kdf_itersalted_s2k(struct sc_context *ctx, const char *hash, const uint8_t *pin, size_t pinlen, const uint8_t *salt, size_t saltlen, uint32_t iterations, uint8_t **out, size_t *outlen)
+{
+	if (!ctx || !hash || !pin || pinlen == 0 || !salt || saltlen != 8 || !out || !outlen) {
+		return SC_ERROR_INVALID_ARGUMENTS;
+	}
+	EVP_MD *md = sc_evp_md(ctx, hash);
+	if (md == NULL) {
+		return SC_ERROR_INVALID_ARGUMENTS;
+	}
+	int r = SC_ERROR_INTERNAL;
+	EVP_MD_CTX *hctx = EVP_MD_CTX_new();
+	if (hctx == NULL) {
+		goto out1;
+	}
+	if (!EVP_DigestInit(hctx, md)) {
+		goto out2;
+	}
+	const int keysize = EVP_MD_size(md);
+	if (keysize <= 0) {
+		goto out2;
+	}
+	uint8_t *outkey = (uint8_t *)sc_mem_secure_alloc(keysize);
+	if (outkey == NULL) {
+		r = SC_ERROR_NOT_ENOUGH_MEMORY;
+		goto out2;
+	}
+	size_t count = iterations;
+	if (count < pinlen + saltlen) {
+		count = pinlen + saltlen;
+	}
+	int ret;
+	while (count >= pinlen + saltlen) {
+		ret = EVP_DigestUpdate(hctx, salt, saltlen);
+		if (!ret) {
+			goto out3;
+		}
+		ret = EVP_DigestUpdate(hctx, pin, pinlen);
+		if (!ret) {
+			goto out3;
+		}
+		count -= pinlen + saltlen;
+	}
+	if (count <= saltlen) {
+		ret = EVP_DigestUpdate(hctx, salt, count);
+	} else {
+		ret = EVP_DigestUpdate(hctx, salt, saltlen);
+		if (!ret) {
+			goto out3;
+		}
+		count -= saltlen;
+		ret = EVP_DigestUpdate(hctx, pin, count);
+	}
+	if (!ret) {
+		goto out3;
+	}
+	unsigned int outsize = keysize;
+	ret = EVP_DigestFinal(hctx, outkey, &outsize);
+	if (ret && outsize == (unsigned int)keysize) {
+		r = SC_SUCCESS;
+	}
+out3:
+	if (r != SC_SUCCESS) {
+		sc_mem_secure_clear_free(outkey, keysize);
+	}
+out2:
+	EVP_MD_CTX_free(hctx);
+out1:
+	sc_evp_md_free(md);
```

Comment:
I would prefer to have just one label for a goto. More jumping makes the code harder to read.

If you initialize the pointers to NULL at the beginning, its safe to call on them the `*_free()`. regardless you passed some point in the code or not.

This also involves declaring the variables at the beginning of the block rather than through the function, which might sounds old-fashined, but very helpful to get the code that is easy to follow.

I would suggest you to do it also in the other functions you did as mixing up declarations and code does not make the code easier to read.